accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+
+This requirement currently only applies when executing as Apple Silicon native code.
+
+This is not expected to impact our customers as it needs to be handled automatically by our run-times.
+
+#### Tightened Signing Requirements
+
+When running on Apple Silicon unsigned binaries will get killed automatically by the kernel. A simple non-obvious message will be printed on the console (`zsh killed: helloWorld`).
+
+To allow a native shared library to be loaded, it will also need to be signed.
+
+When running the equivalent unsigned binary from a Rosetta 2 emulated process the process is not killed. When running a x64 process on Apple Silicon the process was not killed either.
+
+XCode 12 adhoc signs native binaries. These anonymous signature are sufficient to prevent these from being killed (on the machine that compiled them at least.)
+
+I haven't found the documentation exactly how and what changed. These behavior descriptions have been through experimentation and debugging.
+
+### Microsoft One.NET Transition
+
+We are in the middle of a migration from two distinct .NET ecosystems (Mono & .NET Core), to a shared ecosystem, One.NET.
+
+This document was originally written form a pure CoreCLR perspective and was missing a Mono perspective. As Mono feedback is received it will be incorporate to make this a better design.
+
+Any feedback is appreciated.
+
+### New Architecture and .NET Back-porting
+
+Support for a new processor architecture is a significant undertaking.  It is too large an undertaking to consider on our stable release branches.  New architecture development is occurring in .NET 6.
+
+#### CoreCLR Runtime
+
+We do not plan to back-port Apple Silicon support to the current CoreCLR runtime releases .NET 2.1, .NET 3.1, and .NET 5.0.
+
+We have tested the .NET 5 release with Rosetta 2.  Apple has resolved the majority of issues we reported to them.  This means that CoreCLR .NET 2.1 LTS, .NET 3.1 LTS and .NET 5 will mostly just work on Apple Silicon.
+
+## .NET 6 Preview1 CoreCLR User Experience Issues
+
+### Mixed Architecture Side By Side Installs Fail
+
+While the customer can install the macOS Apple Silicon .NET preview build by itself, install this beside the macOS x64 builds is problematic.
+
+We have several possible representative scenarios. Two typical ones are described:
+
+- .NET 3.1 x64 with .NET 6 Apple Silicon. As side effect of not back-porting older releases to the new architecture is that customer will need to deal with multi-architecture side by side installs at least until .NET 3.1 goes out of support. However, this mixed forced architecture problem is transient in nature.
+
+- .NET 6 x64 with .NET 6 Apple Silicon. This scenario is not absolutely critical, but may be required by some customers.
+  - It enables moving legacy Apps with x64 dependent apps to .NET 6.
+  - It enables testing macOS x64 apps under the Rosetta 2 emulator.
```

Comment:
Running apps via `dotnet app` if both `x64` and `arm64` is supported by the `dotnet` - this is currently problematic because there's nothing in the `app.runtimeconfig.json` which would hint the `dotnet`/`hostfxr` which architecture the app needs.
- We would either have to add this information somewhere (`.runtimeconfig.json` is probably the most natural place for this)
- Rely on `app` (executable) instead
  - Currently this is disabled by default for MacOS RID targets (if I remember correctly)  - this fact is not discussed anywhere in this doc either
  - Requires signing - as mentioned above

The universal muxer should default to one architecture if it doesn't know better (probably arm64)

Reply:
> there's nothing in the app.runtimeconfig.json which would hint the dotnet/hostfxr which architecture the app needs.

This is intentional design. We talk about these as [cross-platform apps that can run anywhere](https://docs.microsoft.com/en-us/dotnet/core/deploying/#produce-a-cross-platform-binary).

Reply:
I know of the existence of "portable apps" - that's why the muxer will need a "default" architecture (for those apps). The point was about current state of things where we don't generate `apphost` for apps on MacOS and thus they're executed via `dotnet app.dll`. Those apps may not be portable, but currently there's no way to tell which RID they were built for (if any) and will break in bad ways if run on the wrong architecture (for example if they have native dependencies).

If we solve the signing problem and turn on `apphost` by default for MacOS, then this problem becomes much less important and we can probably continue to ignore it (just like we ignore it on Win x86/x64 today).